lib/fourOhFour.js

```diff
@@ -0,0 +1,190 @@
+          opts.preHandler = opts.preHandler.map(hook => hook.bind(_fastify))
+        } else {
+          opts.preHandler = opts.preHandler.bind(_fastify)
+        }
+      }
+
+      if (opts.preValidation) {
+        if (Array.isArray(opts.preValidation)) {
+          opts.preValidation = opts.preValidation.map(hook => hook.bind(_fastify))
+        } else {
+          opts.preValidation = opts.preValidation.bind(_fastify)
+        }
+      }
+    }
+
+    if (typeof opts === 'function') {
+      handler = opts
+      opts = undefined
+    }
+    opts = opts || {}
+
+    if (handler) {
+      this[kFourOhFourLevelInstance][kCanSetNotFoundHandler] = false
+      handler = handler.bind(this)
+    } else {
+      handler = basic404
+    }
+
+    this.after((notHandledErr, done) => {
+      _setNotFoundHandler.call(this, prefix, opts, handler, avvio, routeHandler, buildMiddie)
+      done(notHandledErr)
+    })
+  }
+
+  function _setNotFoundHandler (prefix, opts, handler, avvio, routeHandler, buildMiddie) {
+    const context = new Context(
+      opts.schema,
+      handler,
+      this[kReply],
+      this[kRequest],
+      this[kContentTypeParser],
+      opts.config || {},
+      this._errorHandler,
+      this[kBodyLimit],
+      this[kLogLevel]
+    )
+
+    avvio.once('preReady', () => {
+      const context = this[kFourOhFourContext]
+
+      const onRequest = this[kHooks].onRequest
+      const preParsing = this[kHooks].preParsing.concat(opts.preParsing || [])
+      const preValidation = this[kHooks].preValidation.concat(opts.preValidation || [])
+      const preSerialization = this[kHooks].preSerialization.concat(opts.preSerialization || [])
+      const preHandler = this[kHooks].preHandler.concat(opts.beforeHandler || opts.preHandler || [])
+      const onSend = this[kHooks].onSend
+      const onError = this[kHooks].onError
+      const onResponse = this[kHooks].onResponse
+
+      context.onRequest = onRequest.length ? onRequest : null
+      context.preParsing = preParsing.length ? preParsing : null
+      context.preValidation = preValidation.length ? preValidation : null
+      context.preSerialization = preSerialization.length ? preSerialization : null
+      context.preHandler = preHandler.length ? preHandler : null
+      context.onSend = onSend.length ? onSend : null
+      context.onError = onError.length ? onError : null
+      context.onResponse = onResponse.length ? onResponse : null
+
+      context._middie = buildMiddie(this[kMiddlewares])
+    })
+
+    if (this[kFourOhFourContext] !== null && prefix === '/') {
+      Object.assign(this[kFourOhFourContext], context) // Replace the default 404 handler
+      return
+    }
+
+    this[kFourOhFourLevelInstance][kFourOhFourContext] = context
+
+    this[kFourOhFourLevelInstance][kFourOhFour].router.all(prefix + (prefix.endsWith('/') ? '*' : '/*'), routeHandler, context)
+    this[kFourOhFourLevelInstance][kFourOhFour].router.all(prefix || '/', routeHandler, context)
```

Comment:
There's only one router, so these two lines can just use the `router` variable in the closure.

```suggestion
    router.all(prefix || '/', routeHandler, context)
```